### ISSUE TITLE
Training with Raw Image data as inputs

### ISSUE THREAD

**Issue by mubaidr** [BrainJS/brain.js#176]
![A GIF or MEME to give some spice of the internet](https://media1.tenor.com/images/09c075a32556feeb23c5943c3b5bd496/tenor.gif?itemid=3320482)

## *What* is wrong?
Training error does not go below 1, and I think this is because my inputs strings have different lengths. 

(currently i am using `base64` encoded image data as input to the network, is this OK?)

## *Where* does it happen?
In the a new brain.recurrent.RNN() when trying to run a net in node.js on my Windows 10

## *How* do we replicate the issue?
Git Source here: https://github.com/mubaidr/image-parsing/blob/master/src/index.js#L45

## *How* important is this (1-5)?
3

## Expected behavior (i.e. solution)
Any suggestions to make input data to be same length for all images or maybe i am doing something wrong? 

## Other Comments
Thanks for the awesome work!

**Comment by jusso-dev** [comment]
I don't have an answer to this but + 1 for the gif

**Comment by mubaidr** [comment]
**UPDATE:** 
I am now using raw data from the image (which is same size across all images) But the error is still in the range 400-500.

Also when using simple neural network error is NAN

**Comment by jusso-dev** [comment]
Neural network will never work with data that isn't numerical. What happens if you one-hot encode your input before feeding it into the network? What does your error rate look like?

**Comment by robertleeplummerjr** [comment]
You could try something like this: https://jsfiddle.net/8kpq58bu/117/
Here we encode the 256 colors like: `color / 255`

**Comment by mubaidr** [comment]
`Neural network will never work with data that isn't numerical. What happens if you one-hot encode your input before feeding it into the network? What does your error rate look like?`

Well, I tried using RAW data from the image, which is essentially an array of integer values of each pixel(gray scaled and threshold applied, so values are either 0 or 1). Still error rate fluctuates between 200-400. 

`You could try something like this: https://jsfiddle.net/8kpq58bu/117/
Here we encode the 256 colors like: color / 255`

Thanks I will try this technique and see if it improves anything. 

**UPDATE:** 
What if I try to compress image data to some string format by representing series of pixel using Huffman style encoding? For example, if there are 31 zeros in a sequence the i replace them with `0-31` in my resulting string and similarly for 1. 

e.g. For this data: `00011000000001111000001` result should be: `0-3,1-2,0-8,1-4,0-5,1-1`

Sample Images I am trying to train: 
https://github.com/mubaidr/image-parsing/tree/master/data/sample

Thanks everyone, I am learning new thing each time you guide me.

**Comment by mubaidr** [comment]
**UPDATE:**

**What i am trying to do:** 
![untitled](https://user-images.githubusercontent.com/2222702/38042319-7e392b60-32cd-11e8-9184-f0954c1509e2.png)


Ok So I have tried using following input/output sets for training so far and nothing works, training error **approaches** almost 1: 

- Image data as base64
```js
trainingData={input : ['base64-image-characterset-genrated-using-fs-module'],
output: ['a']}
// and
trainingData={input : ['base64-image-characterset-genrated-using-fs-module'],
output: [{option: 'a'}]}
// Above two training sample generate different training error! wot!
```

- Image Raw pixel data
```js
trainingData={input : [[/*integer array of image raw data*/]], // Also tried {input : [...[/*integer array of image raw data*/]],
output: ['a']} // also tried output: [{option: 'a'}]}
```

- Image data as monochrome/grayscale with threshold (training samples tried same as above)

**Observations**

- In all cases Training error never goes below.
- I have tried using 3 types of neural networks and different activation techniques too. 

**Questions:** 

- Why does changing output format effects error rate? 
- What would the best format for image data to be trained?

**Source**
All the source for this node.js application is available here: 
https://github.com/mubaidr/image-parsing

**Comment by robertleeplummerjr** [comment]
@jusso-dev You can feed representations of data into a neural network, specifically the recurrent neural network, like the [example of the children's book](https://github.com/BrainJS/brain.js/blob/650c322a16b6c76b8adc374f3a9afb83ca6e50ee/examples/childrens-book.js#L3)

Also, I see that is what is happening [here](https://github.com/mubaidr/image-parsing/blob/3bfab42ed25fe8c3fb867eb273fafaaeda9dffe7/src/index.js#L9), neat!

@mubaidr `brain.recurrent.RNN` will likely not give you good results, because it is not sophisticated enough to retain what data you are trying to send in.  It is a vanilla recurrent neural network and thus suffers from exploding or disappearing gradients. (here is a link to a fantastic book I love refering too that explains it in practical terms: https://leonardoaraujosantos.gitbooks.io/artificial-inteligence/content/recurrent_neural_networks.html#long-short-term-memory-networkslstm, if you like the book, be sure to share it and star it here: https://www.gitbook.com/book/leonardoaraujosantos/artificial-inteligence/details)

There are many ways to break down this problem, but you could try and use the `brain.recurrent.LSTM` class, that will likely be able to help.

If you were feeling a little adventurous, and would like to help out with the new layer composition and gpu performance features, you could venture into using the [branch](https://github.com/BrainJS/brain.js/blob/nn-gpu-layers) and try out the new [`FeedForward` class](https://github.com/BrainJS/brain.js/blob/nn-gpu-layers/src/feed-forward.js) and perhaps use a [convolution](https://github.com/BrainJS/brain.js/blob/nn-gpu-layers/src/layer/convolution.js) layer which is better at compressing the inputs into patters that can more easily be classified.  More on convolutions [here](https://leonardoaraujosantos.gitbooks.io/artificial-inteligence/content/convolutional_neural_networks.html).

**Comment by mubaidr** [comment]
I have already tried `LSTM` but no result. One of my main concern here is that if my input/output format is correct in this application? @robertleeplummerjr 

Thanks for book reference, I will definitely try to read and learn. Also, gonna try new branch.

**Comment by mubaidr** [comment]
**UPDATE:**

- NN
```
Preparing training data...
[ { input: [ 1, 0, 0, 0 ], output: [ 'a' ] },
  { input: [ 0, 1, 0, 0 ], output: [ 'b' ] },
  { input: [ 0, 0, 1, 0 ], output: [ 'c' ] },
  { input: [ 0, 0, 0, 1 ], output: [ 'd' ] } ]

Training started...


Training finished in 0.014s with error: NaN
```

- RNN
```
Preparing training data...
[ { input: [ 1, 0, 0, 0 ], output: [ 'a' ] },
  { input: [ 0, 1, 0, 0 ], output: [ 'b' ] },
  { input: [ 0, 0, 1, 0 ], output: [ 'c' ] },
  { input: [ 0, 0, 0, 1 ], output: [ 'd' ] } ]

Training started...


Training finished in 0.674s with error: 1.3745829649011243
```

- LSTM
```
Preparing training data...
[ { input: [ 1, 0, 0, 0 ], output: [ 'a' ] },
  { input: [ 0, 1, 0, 0 ], output: [ 'b' ] },
  { input: [ 0, 0, 1, 0 ], output: [ 'c' ] },
  { input: [ 0, 0, 0, 1 ], output: [ 'd' ] } ]

Training started...


Training finished in 2.623s with error: 1.2690826830737323
```

**Comment by mubaidr** [comment]
**UPDATE:**

Thanks everybody for the guidance, I have been able to solve this problem using simple Neural-networks and using One Dimensional data fro the image. 

I issue was related to (mainly) input/output format. Now the training data sample looks like this:
```js
[
{
input: [/*Array from monochrome Image (0s and 1s only)*/],
output: {
a:1
// OR b: 1 depending on the data
}
},
....
]
```

**So in short, encoding my image data into simpler integer array solved this problem.**

**Comment by jusso-dev** [comment]
Feel free to share the finished code

**Comment by mubaidr** [comment]
Everything is opensource under this repository, including sample images, test script: https://github.com/mubaidr/image-parsing

👍 

Preparing Image data: https://github.com/mubaidr/image-parsing/blob/master/src/index.js#L30
Input to NN: https://github.com/mubaidr/image-parsing/blob/master/src/index.js#L115

**Comment by jusso-dev** [comment]
Awesome thanks 👌😉

**Comment by robertleeplummerjr** [comment]
Fantastic.  Mind if we use it as an example?

**Comment by mubaidr** [comment]
@robertleeplummerjr Sure, please go ahead! Let me know if I can help with anything! 🍻

Or I can host this example on some platform and and you can add live link to it? 🤔

**Comment by KokoDoko** [comment]
Does this example only work with monochrome images? Wouldn't it be possbible to use r g b values from 0 to 1 in a neural network? Color seems quite important for image recognition.
```
// a 2x2 pixel image
const image = [0.2, 0.23, 0.48, 0.2, 0.23, 0.48,0.2, 0.23, 0.48,0.2, 0.23, 0.48]
```

**Comment by mubaidr** [comment]
It should work with color data too. In my case images were simpler and color information was not needed.

**Comment by robertleeplummerjr** [comment]
>Wouldn't it be possbible to use r g b values from 0 to 1 in a neural network?

In gpu.js (which is the engine for brain.js v2) we've built it so you can feed images in _directly_: https://github.com/gpujs/gpu.js#supported-input-types

```js
const myFunc = gpu.createKernel(function(image) {
    const pixel = image[this.thread.y][this.thread.x];
    this.color(pixel[0], pixel[1], pixel[2], pixel[3]);
})
  .setGraphical(true)
  .setOutput([100]);

const image = document.createElement('img');
image.src = 'my/image/source.png';
image.onload = () => {
  myFunc(image);
  // Result: colorful image
};
```

If we brought this functionality into v2 of brain.js, it would likely be:
1. after the release of brain.js v2
2. after the release of gpu.js v2 with OpenCL support

I've personally hoped for it for some time, and I think, although challenging, it would be terribly useful.  However, we'd need to plan for how that api would look and ensure it'd be useful for real world scenarios.

_However_, @mubaidr raises a good point:

>In my case images were simpler and color information was not needed.

Most scenarios really don't need color data.  But I figure, if you are going to make something useful...

![](https://media2.giphy.com/media/U9RNGy5TeAX3W/giphy.gif)

Why not make it easy?

(Caution: possibly overly simplified use case)

```js
net.run(<img />);

// <img /> = document.createElement('img');
```

**Comment by robertleeplummerjr** [comment]
@mubaidr https://github.com/mubaidr/image-parsing is down.

**Comment by maxall41** [comment]
is there a working link to  https://github.com/mubaidr/image-parsing?

**Comment by mubaidr** [comment]
Repository is available now: https://github.com/mubaidr/image-parsing 

Prepare image data: https://github.com/mubaidr/image-parsing/blob/master/src/utilities/convertToBitArray.ts

I will try to publish a simpler code pen.

**Comment by TheEpicFace007** [comment]
> Repository is available now: https://github.com/mubaidr/image-parsing
> 
> Prepare image data: https://github.com/mubaidr/image-parsing/blob/master/src/utilities/convertToBitArray.ts
> 
> I will try to publish a simpler code pen.

the repo is no longer availaible. what do i need to do
